src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/core/CHIPTLV.h"
+#include "lib/core/CHIPTLVTags.h"
+#include "lib/core/CHIPTLVTypes.h"
+#include "protocols/interaction_model/Constants.h"
+#include "system/SystemPacketBuffer.h"
+#include "system/TLVPacketBufferBackingStore.h"
+#include <app/BufferedReadCallback.h>
+#include <app/InteractionModelEngine.h>
+#include <lib/support/ScopedBuffer.h>
+
+namespace chip {
+namespace app {
+
+void BufferedReadCallback::OnReportBegin(const ReadClient * apReadClient)
+{
+    mCallback.OnReportBegin(apReadClient);
+}
+
+void BufferedReadCallback::OnReportEnd(const ReadClient * apReadClient)
+{
+    CHIP_ERROR err = DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCallback.OnError(apReadClient, err);
+    }
+
+    mCallback.OnReportEnd(apReadClient);
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(TLV::ScopedBufferTLVReader & aReader)
+{
+    TLV::TLVType outerType;
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+
+    //
+    // To generate the final reconstituted list, we need to allocate a contiguous
+    // buffer than can hold the entirety of its contents. To do so, we need to figure out
+    // how big a buffer to allocate. This requires walking the buffered list items and computing their TLV sizes,
+    // sum theming all up and adding a bit of slop to account for the TLV array the list elements will go into.
+    //
+    // The alternative was to use a PacketBufferTLVWriter backed by chained packet buffers to
+    // write out the list - this would have removed the need for this first pass. However,
+    // we cannot actually back a TLVReader with a chained buffer since that violates the ability
+    // for us to create readers off-of readers. Each reader would assume exclusive ownership of the chained
+    // buffer and mutate the state within TLVPacketBufferBackingStore, preventing shared use.
+    //
+    // To avoid that, a single contiguous buffer is the best likely approach for now.
+    //
+    uint32_t totalBufSize = 0;
+    for (size_t i = 0; i < mBufferedList.size(); i++)
+    {
+        auto bufHandle = mBufferedList[i].Retain();
```

Comment:
Why do we need the `Retain()` here?  I'd think we can just `totalBufSize += mBufferedList[[i]->TotalLength();`